src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,316 @@
+#include <Common/typeid_cast.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Interpreters/ArithmeticOperationsInAgrFuncOptimize.h>
+#include <IO/WriteHelpers.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int UNEXPECTED_AST_STRUCTURE;
+}
+/// scalar values from the first level
+std::pair<ASTs, ASTs> tryToCatchConst(std::string & name, ASTs & argss)
+{
+    ASTs const_num;
+    ASTs not_const;
+
+    for (const auto &ar: argss)
+    {
+        if (const auto * literal = ar->as<ASTLiteral>())
+        {
+            if (literal->value.getType() == Field::Types::Int64 ||
+            literal->value.getType() == Field::Types::UInt64 ||
+            literal->value.getType() == Field::Types::Int128 ||
+            literal->value.getType() == Field::Types::UInt128)
+            {
+                const_num.push_back(ar);
+            }
+            else
+            {
+                not_const.push_back(ar);
+            }
+        }
+        else
+        {
+            not_const.push_back(ar);
+        }
+    }
+
+    if ((name == "plus" || name == "multiply") && const_num.size() + not_const.size() != 2)
+    {
+        throw Exception("Wrong number of arguments for function 'plus' or 'multiply' (" + toString(const_num.size() + not_const.size()) + " instead of 2)",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+    }
+
+    return {const_num, not_const};
+}
+
+std::pair<ASTs, ASTs> findAllConsts(ASTFunction * func_node, std::string & inter_func_name)
+{
+    if (!func_node->arguments)
+    {
+        return {};
+    }
+    else if (!(func_node->arguments->children[0]->as<ASTFunction>()))
+    {
+        auto arg = func_node->arguments->children;
+        std::pair<ASTs, ASTs> it = tryToCatchConst(func_node->name, arg);
+        return it;
+    }
+    else if (inter_func_name != func_node->arguments->children[0]->as<ASTFunction>()->name)
+    {
+        return {};
+    }
+    else
```

Comment:
Bad code style. Remove all `else` causes here. Main `if` branch returns without condition. So if we are still in this function we are in `else`.